mysqld_exporter.go

```diff
@@ -164,24 +238,76 @@ func (e *Exporter) scrape(scrapes chan<- []string) {
 	e.duration.Set(float64(time.Now().UnixNano()-now) / 1000000000)
 }
 
+func (e *Exporter) setGenericMetric(name string, value float64) {
+	if _, ok := e.metrics[name]; !ok {
+		e.metrics[name] = prometheus.NewUntyped(prometheus.UntypedOpts{
+			Namespace: namespace,
+			Name:      name,
+		})
+	}
+
+	e.metrics[name].Set(value)
+}
+
 func (e *Exporter) setMetrics(scrapes <-chan []string) {
-	for row := range scrapes {
+	var comRegex = regexp.MustCompile(`^com_(.*)$`)
+	var connectionErrorRegex = regexp.MustCompile(`^connection_errors_(.*)$`)
+	var handlerRegex = regexp.MustCompile(`^handler_(.*)$`)
+	var innodbPageRegex = regexp.MustCompile(`^innodb_pages_(.*)$`)
+	var innodbRowRegex = regexp.MustCompile(`^innodb_rows_(.*)$`)
+	var performanceSchemaRegex = regexp.MustCompile(`^performance_schema_(.*)$`)
+	var sslConnectionRegex = regexp.MustCompile(`^ssl_(.*_renegotiates|accepts|client_connects|connect_renegotiates|finished_.*|)$`)
+	var sslSessionCacheRegex = regexp.MustCompile(`^ssl_session_cache_(hits|misses|overflows|timeouts)$`)
+	var threadRegex = regexp.MustCompile(`^threads_(.*)$`)
 
+	for row := range scrapes {
 		name := strings.ToLower(row[0])
-		value, err := strconv.ParseInt(row[1], 10, 64)
+		value, err := strconv.ParseFloat(row[1], 64)
 		if err != nil {
 			// convert/serve text values here ?
 			continue
 		}
 
-		if _, ok := e.metrics[name]; !ok {
-			e.metrics[name] = prometheus.NewGauge(prometheus.GaugeOpts{
-				Namespace: namespace,
-				Name:      name,
-			})
+		command := comRegex.FindStringSubmatch(name)
+		connectionError := connectionErrorRegex.FindStringSubmatch(name)
+		handler := handlerRegex.FindStringSubmatch(name)
+		innodbPage := innodbPageRegex.FindStringSubmatch(name)
+		innodbRow := innodbRowRegex.FindStringSubmatch(name)
+		performanceSchema := performanceSchemaRegex.FindStringSubmatch(name)
+		sslConnection := sslConnectionRegex.FindStringSubmatch(name)
+		sslSessionCache := sslSessionCacheRegex.FindStringSubmatch(name)
+		thread := threadRegex.FindStringSubmatch(name)
+
+		switch {
+		case len(command) > 0:
+			e.commands.With(prometheus.Labels{"command": command[1]}).Set(value)
+		case len(connectionError) > 0:
+			e.connectionErrors.With(prometheus.Labels{"error": connectionError[1]}).Set(value)
+		case len(handler) > 0:
+			e.handlers.With(prometheus.Labels{"handler": handler[1]}).Set(value)
+		case len(innodbPage) > 0:
+			if innodbPage[1] == "created" {
```

Comment:
Special cases make me feel uncomfortable in an exporter, this feels like something that might require maintenance over time as new statuses relating to innodb come along